staging/src/k8s.io/apiserver/pkg/server/storage/storage_factory.go

```diff
@@ -263,6 +282,101 @@ func (s *DefaultStorageFactory) NewConfig(groupResource schema.GroupResource) (*
 	return storageConfig.ForResource(groupResource), nil
 }
 
+type removedInterface interface {
+	APILifecycleRemoved() (major, minor int)
+}
+
+// Object interface generated from "k8s:prerelease-lifecycle-gen:introduced" tags in types.go.
+type introducedInterface interface {
+	APILifecycleIntroduced() (major, minor int)
+}
+
+func emulatedStorageVersion(binaryVersionOfResource schema.GroupVersion, example runtime.Object, effectiveVersion version.EffectiveVersion) (schema.GroupVersion, error) {
+	if example == nil || effectiveVersion == nil {
+		return binaryVersionOfResource, nil
+	}
+
+	// Look up example in scheme to find all objects of the same Group-Kind
+	// Use the highest priority version for that group-kind whose lifecycle window
+	// includes the current emulation version.
+	// If no version is found, use the binary version
+	// (in this case the API should be disabled anyway)
+
+	//!TODO: Is there a better scheme to use here? Does this include
+	// aggregator scheme/apiextensions scheme? What about for aggregated apiservers?
+	// May have to augment constructor with list of schemes to try...
+	sch := legacyscheme.Scheme
+
+	gvks, _, err := sch.ObjectKinds(example)
+	if err != nil {
+		return schema.GroupVersion{}, err
+	} else if len(gvks) == 0 {
+		// Probably shouldn't happen if err is non-nil
+		return schema.GroupVersion{}, fmt.Errorf("Object %T has no GVKs registered in scheme", example)
+	}
+
+	versions := sch.VersionsForGroupKind(schema.GroupKind{Group: gvks[0].Group, Kind: gvks[0].Kind})
+
+	//!TODO: what emulation version to use? Is it dependent upon component
+	// which depends upon source of the object?
+	// Are we mapping component version to a global k8s version?
+	//
+	// Purposefully using MinCompatibilityVersion rather than EmulationVersion
+	// for storage version determinations because we want the persisted resource
+	// to be compatible with MinCompatibilityVersion, not EmulationVersion.
+	compatibilityVersion := effectiveVersion.MinCompatibilityVersion()
+
+	for _, gv := range versions {
+		if gv.Version == runtime.APIVersionInternal {
+			continue
+		}
+
+		gvk := schema.GroupVersionKind{
+			Group:   gv.Group,
+			Version: gv.Version,
+			Kind:    gvks[0].Kind,
+		}
+
+		exampleOfGVK, err := sch.New(gvk)
+		if err != nil {
+			return schema.GroupVersion{}, err
+		}
+
+		// If it was introduced after current compatibility version, don't use it
+		introduced, hasIntroduced := exampleOfGVK.(introducedInterface)
+		if hasIntroduced {
+			majorIntroduced, minorIntroduced := introduced.APILifecycleIntroduced()
+			introducedVer := apimachineryversion.MajorMinor(uint(majorIntroduced), uint(minorIntroduced))
+			if introducedVer.GreaterThan(compatibilityVersion) {
+				continue
+			}
+		}
+
+		// If it was removed before current compatibility version, don't use it
+		removed, hasRemoved := exampleOfGVK.(removedInterface)
+		if hasRemoved {
+			majorRemoved, minorRemoved := removed.APILifecycleRemoved()
```

Comment:
APILifecycleRemoved is when we stop *serving* a version, but we still understand how to decode / convert the version, so I don't think you need to exclude "removed" versions